addons/website_sale_loyalty/controllers/main.py

```diff
@@ -21,8 +21,11 @@ def pricelist(self, promo, **post):
             reward_successfully_applied = True
             if len(coupon_status) == 1:
                 coupon, rewards = next(iter(coupon_status.items()))
-                if len(rewards) == 1 and not rewards.multi_product:
-                    reward_successfully_applied = self._apply_reward(order, rewards, coupon)
+                if len(rewards) == 1 and (product or not rewards.multi_product):
+                    request.env.context = dict(request.env.context, product=product)
```

Comment:
Mhm, you're only using the key `product_id` below, not `product`.
Please make sure that this part is tested too.
Also, why don't you use the `update_context` method on the request, as proposed in my comment?

Reply:
Oh ok, i added the argument in pricelist because when the task was for master the claim_reward function would call `pricelist`,  [here](https://github.com/odoo/odoo/blob/master/addons/website_sale_loyalty/controllers/main.py#L99) .
so there was a particular case where claim would not work
since that's not the case in 16 i will remove it.